src/filter/mod.rs

```diff
@@ -446,14 +447,22 @@ fn apply2<'a>(
                 require_literal_separator: true,
                 require_literal_leading_dot: true,
             };
-            tree::remove_pred(
+            let res = tree::remove_pred(
                 &repo,
                 "",
                 tree.id(),
                 &|path, isblob| isblob && (pattern.matches_path_with(&path, options)),
                 git2::Oid::zero(),
                 &mut std::collections::HashMap::new(),
-            )
+            );
+            if let Some(warns) = warnings {
```

Comment:
I think it should be possible to do it just once for any filter, basically it's when apply() return empty_tree and then report "filter does not match anything"

Reply:
Maybe even doing it on a flattened compose. Then it would not pollute the rest of the code.